### PR TITLE
Added adaptive text editor theme

### PR DESCRIPTION
**Pull request by djrm**
Added text editor theme that adapts according to the current theme.

**Edit**: also works with shader editor.

screenshot from 2017-09-14 19-39-29
screenshot from 2017-09-14 19-39-46

### CODE REVIEW

**Comment by Paulb23** [comment]
You might want to add something similar to `shader_editor_plugin.cpp`

**Comment by Paulb23** [comment]
You will also need to add a check to`save_text_editor_theme` and `save_text_editor_theme_as` in EditorSettings for "adaptive".

File: scene/gui/text_edit.cpp [hunk 1]
```diff
@@ -729,11 +729,6 @@ void TextEdit::_notification(int p_what) {
 					VisualServer::get_singleton()->canvas_item_add_rect(ci, Rect2(xmargin_beg, ofs_y, xmargin_end - xmargin_beg, get_row_height()), cache.mark_color);
 				}
 
-				if (text.is_breakpoint(line)) {
-
-					VisualServer::get_singleton()->canvas_item_add_rect(ci, Rect2(xmargin_beg, ofs_y, xmargin_end - xmargin_beg, get_row_height()), cache.breakpoint_color);
-				}
-
```
Comment (Paulb23) [inline scene/gui/text_edit.cpp]: Can't really remove this, as if the user has `show_breakpoint_gutter` to false they will have no visual clue to where breakpoints are.

File: editor/plugins/script_text_editor.cpp [hunk 2]
```diff
@@ -209,7 +290,12 @@ void ScriptTextEditor::_notification(int p_what) {
 	if (p_what == NOTIFICATION_READY) {
 
 		//emit_signal("name_changed");
+		_load_theme_settings();
 	}
+	// if (p_what == EDITOR_SETTINGS_CHANGED) {
+
+	// 	_load_theme_settings();
+	// }
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp]: Not to sure @akien-mga will agree with this :P

**Comment by djrm** [comment]
@Paulb23 the adaptive theme is not meant to be modified, thats why i left the default theme, that you can modify, also yeah, will see about shader editor.

File: scene/gui/text_edit.cpp (hunk 1 above)
Comment (djrm) [reply scene/gui/text_edit.cpp]: ok, then will change that so it does not look so intrusive.

File: editor/plugins/script_text_editor.cpp (hunk 2 above)
Comment (djrm) [reply editor/plugins/script_text_editor.cpp]: forgot about it.

**Comment by djrm** [comment]
@Paulb23 oh i see what you mean.

File: editor/editor_settings.cpp [hunk 3]
```diff
@@ -599,8 +599,8 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	hints["global/default_project_export_path"] = PropertyInfo(Variant::STRING, "global/default_project_export_path", PROPERTY_HINT_GLOBAL_DIR);
 	set("interface/show_script_in_scene_tabs", false);
 
-	set("text_editor/theme/color_theme", "Default");
-	hints["text_editor/theme/color_theme"] = PropertyInfo(Variant::STRING, "text_editor/theme/color_theme", PROPERTY_HINT_ENUM, "Default");
+	set("text_editor/theme/color_theme", "Adaptive");
+	hints["text_editor/theme/color_theme"] = PropertyInfo(Variant::STRING, "text_editor/theme/color_theme", PROPERTY_HINT_ENUM, "Adaptive,Custom");
```
Comment (Paulb23) [inline editor/editor_settings.cpp]: Might want to keep this one as "Adaptive,Default" to be safe unless you want to change everywhere to use "Custom" (it gets overwritten anyway).

File: editor/editor_themes.cpp [hunk 4]
```diff
@@ -782,6 +783,79 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
...
+	const Color te_background_color = Color(0, 0, 0, 0);
+	const Color completion_background_color = base_color;
+	const Color completion_selected_color = alpha1;
+	const Color completion_existing_color = alpha2;
+	const Color completion_scroll_color = alpha1;
+	const Color completion_font_color = font_color;
+	const Color text_color = font_color;
+	const Color line_number_color = dim_color;
+	const Color caret_color = mono_color;
+	// caret_background_color  // what is this for?
```
Comment (Paulb23) [inline editor/editor_themes.cpp]: This is the colour of the text behind the caret when `block_caret` is used.

File: editor/editor_settings.cpp (hunk 3 above)
Comment (djrm) [reply editor/editor_settings.cpp]: oh true.

File: editor/editor_themes.cpp (hunk 4 above)
Comment (djrm) [reply editor/editor_themes.cpp]: good to know

File: editor/plugins/script_text_editor.cpp [hunk 5]
```diff
@@ -76,34 +76,97 @@ void ScriptTextEditor::_load_theme_settings() {
 	text_edit->clear_colors();
 
 	/* keyword color */
```
Comment (Paulb23) [inline editor/plugins/script_text_editor.cpp]: Could remove this comment as it no longer makes scene, same in the shade editor.

File: scene/gui/text_edit.cpp [hunk 6]
```diff
@@ -729,15 +730,14 @@ void TextEdit::_notification(int p_what) {
...
 				if (line == cursor.line) {
 					VisualServer::get_singleton()->canvas_item_add_rect(ci, Rect2(0, ofs_y, xmargin_end, get_row_height()), cache.current_line_color);
 				}
 
+				if (text.is_breakpoint(line) && !draw_breakpoint_gutter) {
+					VisualServer::get_singleton()->canvas_item_add_rect(ci, Rect2(xmargin_beg, ofs_y + get_row_height() - EDSCALE, xmargin_end - xmargin_beg, EDSCALE), cache.breakpoint_color);
+				}
+
```
Comment (Paulb23) [inline scene/gui/text_edit.cpp]: Screenshot? :)

Comment (djrm) [reply scene/gui/text_edit.cpp]: there you go
screenshot from 2017-09-15 16-05-03

**Comment by Paulb23** [comment]
Looks good to me!

**Comment by Noshyaar** [comment]
Hmm... this PR segfaults the editor somehow.

```
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /usr/lib/libc.so.6(+0x34920) [0x7fb527545920] (??:0)
[2] ScriptTextEditor::_load_theme_settings() (/home/noshyaar/git/godot/0/editor/plugins/script_text_editor.cpp:172 (discriminator 11))
[3] ScriptTextEditor::_notification(int) (/home/noshyaar/git/godot/0/editor/plugins/script_text_editor.cpp:271)
[4] ScriptTextEditor::_notificationv(int, bool) (/home/noshyaar/git/godot/0/editor/plugins/script_text_editor.h:38 (discriminator 14))
[5] Object::notification(int, bool) (/home/noshyaar/git/godot/0/core/object.cpp:851)
[6] Node::_propagate_ready() (/home/noshyaar/git/godot/0/scene/main/node.cpp:180)
[7] Node::_set_tree(SceneTree*) (/home/noshyaar/git/godot/0/scene/main/node.cpp:2529)
[8] Node::_add_child_nocheck(Node*, StringName const&) (/home/noshyaar/git/godot/0/scene/main/node.cpp:1308)
[9] Node::add_child(Node*, bool) (/home/noshyaar/git/godot/0/scene/main/node.cpp:1338 (discriminator 2))
[10] ScriptEditor::edit(Ref<Script> const&, int, int, bool) (/home/noshyaar/git/godot/0/editor/plugins/script_editor_plugin.cpp:1687 (discriminator 2))
[11] ScriptEditor::edit(Ref<Script> const&, bool) (/home/noshyaar/git/godot/0/editor/plugins/script_editor_plugin.h:351)
[12] ScriptEditor::set_scene_root_script(Ref<Script>) (/home/noshyaar/git/godot/0/editor/plugins/script_editor_plugin.cpp:2147)
[13] EditorNode::_set_main_scene_state(Dictionary, Node*) (/home/noshyaar/git/godot/0/editor/editor_node.cpp:2839 (discriminator 2))
[14] MethodBind2<Dictionary, Node*>::call(Object*, Variant const**, int, Variant::CallError&) (/home/noshyaar/git/godot/0/core/method_bind.gen.inc:1428 (discriminator 19))
[15] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/noshyaar/git/godot/0/core/object.cpp:839 (discriminator 1))
[16] MessageQueue::_call_function(Object*, StringName const&, Variant const*, int, bool) (/home/noshyaar/git/godot/0/core/message_queue.cpp:259)
[17] MessageQueue::flush() (/home/noshyaar/git/godot/0/core/message_queue.cpp:306)
[18] SceneTree::idle(float) (/home/noshyaar/git/godot/0/scene/main/scene_tree.cpp:501)
[19] Main::iteration() (/home/noshyaar/git/godot/0/main/main.cpp:1632)
[20] OS_X11::run() (/home/noshyaar/git/godot/0/platform/x11/os_x11.cpp:2127)
[21] /home/noshyaar/git/godot/0/bin/godot.x11.tools.64(main+0xd5) [0xd5e12f] (/home/noshyaar/git/godot/0/platform/x11/godot_x11.cpp:55)
[22] /usr/lib/libc.so.6(__libc_start_main+0xea) [0x7fb527531f6a] (??:0)
[23] /home/noshyaar/git/godot/0/bin/godot.x11.tools.64(_start+0x2a) [0xd5df9a] (??:?)
-- END OF BACKTRACE --                                                                    
```

**Comment by Geequlim** [comment]
Why force the docmentation page to use this theme ?

**Comment by Geequlim** [comment]
And it is still dark background with the documentation page with light editor theme.

**Comment by djrm** [comment]
@Geequlim it does not, also thats another issue.

**Comment by n-pigeon** [comment]
I would saturate some colors a bit more.
For example `Input` has a light cold green tint compared to warm grey of text above and it is difference that is really hard to notice.
If I have hard time to distinguish difference in theming on them, for some people it might be plain impossible.
Same with `lerp() . += var ` they should pop a tad more.

**Comment by djrm** [comment]
@n-pigeon my reasoning behind the colors is the following,
Now the colors means relevance, for example a Singleton is not as important as a built-in type, even though both are types (and thats why they are green), usually you create your scripts by calling multiple functions, that why they require more atention, since most of the logic is descibed by function calls, the reserved words does not need too much attention, things like var, but i thing flow keywords should get their own color (like if, match, for), so thats the reasoning behind, make relevant parts of code more visible, this experiment might need more work but in theory i think this should be the goal, of course i could have made a vibrant theme with lots color but, thats not the purpose.

**Comment by n-pigeon** [comment]
@djrm What I tried to portrait is that this color scheme is so uniform that changes in colors are almost none. If they are in different color I assume it's for some purpose, but with so small changes they are almost like they would have no differences.
I have hard time to see those differences some ppl with worse eyes will see none. Thats what I wanted to point out :)

**Comment by djrm** [comment]
@n-pigeon thats the thing they are similar because they actually are, so Vector3 is a type the same as CollisionShape, buy let me try on a worst monitor to see how it behaves.

**Pull request merged**